src/thermo/PengRobinsonMFTP.cpp

```diff
@@ -0,0 +1,1321 @@
+		for (size_t i = 0; i < m_kk; i++) {
+				size_t counter = i + m_kk * i;
+				// Calculate first and double derivatives of alpha for individual species
+				Tc = speciesCritTemperature(a_vec_Curr_[counter], b_vec_Curr_[i]);
+				sqtTr = sqrt(temperature() / Tc); //we need species critical temperature
+				coeff1 = 1 / (Tc*sqtTr);
+				coeff2 = sqtTr - 1;
+				k = kappa_vec_[i];
+				dalphadT_vec_Curr_[i] = coeff1 *(k* k*coeff2 - k);
+		}
+		//Calculate mixture derivative
+		for (size_t i = 0; i < m_kk; i++) {
+				size_t counter1 = i + m_kk * i;
+				for (size_t j = 0; j < m_kk; j++) {
+						size_t counter2 = j * m_kk + j;
+						temp = 0.5 * sqrt((a_vec_Curr_[counter1] * a_vec_Curr_[counter2]) / (alpha_vec_Curr_[i] * alpha_vec_Curr_[j]));
+						daAlphadT += moleFractions_[i] * moleFractions_[j] * temp
+							* (dalphadT_vec_Curr_[j] * alpha_vec_Curr_[i] + dalphadT_vec_Curr_[i] * alpha_vec_Curr_[j]);
+				}
+		}
+		return daAlphadT;
+}
+
+doublereal PengRobinsonMFTP::d2aAlpha_dT2() const
+{
+		doublereal daAlphadT = 0.0, temp, fac1, fac2, alphaij, alphai, alphaj, d2aAlphadT2 = 0.0, num;
+		double k;
+		double sqt_Tr = sqrt(temperature() / critTemperature()); //we need species critical temperature
+		double coeff1 = 1 / (critTemperature()*critTemperature()*sqt_Tr);
+		double coeff2 = sqt_Tr - 1;
+		for (size_t i = 0; i < m_kk; i++) {
+				// Calculate individual dAlpha_dTi
+				size_t counter = i + m_kk * i;
+				k = kappa_vec_[i];
+				dalphadT_vec_Curr_[i] = coeff1 *(k* k*coeff2 - k);
+				d2alphadT2_[i] = (k*k + k) * coeff1 / (2 * sqt_Tr*sqt_Tr);
+		}
+
+		//Calculate mixture derivative
+		for (size_t i = 0; i < m_kk; i++) {
+				size_t counter1 = i + m_kk * i;
+				alphai = alpha_vec_Curr_[i];
+				for (size_t j = 0; j < m_kk; j++) {
+						size_t counter2 = j + m_kk * j;
+						alphaj = alpha_vec_Curr_[j];
+						alphaij = alphai * alphaj;
+						temp = 0.5 * sqrt((a_vec_Curr_[counter1] * a_vec_Curr_[counter2]) / (alphaij));
+						num = (dalphadT_vec_Curr_[j] * alphai + dalphadT_vec_Curr_[i] * alphaj);
+						fac1 = -(0.5 / alphaij)*num*num;
+						fac2 = alphaj * d2alphadT2_[counter1] + alphai *d2alphadT2_[counter2] + 2 * dalphadT_vec_Curr_[i] * dalphadT_vec_Curr_[j];
+						d2aAlphadT2 += moleFractions_[i] * moleFractions_[j] * temp *(fac1 + fac2);
+				}
+		}
+		return d2aAlphadT2;
+}
+
+void PengRobinsonMFTP::calcCriticalConditions(doublereal a, doublereal b,
+        doublereal& pc, doublereal& tc, doublereal& vc) const
+{
+    if (b <= 0.0) {
+        tc = 1000000.;
+        pc = 1.0E13;
+        vc = omega_vc * GasConstant * tc / pc;
+        return;
+    }
+    if (a <= 0.0) {
+        tc = 0.0;
+        pc = 0.0;
+        vc = 2.0 * b;
+        return;
+    }
+    tc = a * omega_b / (b * omega_a * GasConstant);
+    pc = omega_b * GasConstant * tc / b;
+    vc = omega_vc * GasConstant * tc / pc;
+}
+
+int PengRobinsonMFTP::NicholsSolve(double TKelvin, double pres, doublereal a, doublereal b, doublereal aAlpha,
```

Comment:
This implementation is based on Nickall's method (www.nickalls.org/dick/papers/maths/cubic1993.pdf). I will add the reference in the header file.

Reply:
Can you please give the official reference (which is listed in that PDF as The Mathematical Gazette(1993); 77(November), 354–359), and link to JSTOR (https://www.jstor.org/stable/3619777) instead of to this person's website? I'm afraid this link to `www.nickalls.org` will go dead.

Reply:
The link and the reference has been updated.